rustls/src/msgs/message.rs

```diff
@@ -539,3 +539,257 @@ impl BorrowedPlainMessage for BorrowedPlainMessageOutbound<'_> {
+                        size += chunk.len();
+                    }
+                }
+            }
+        }
+    }
+
+    /// split self in two, around an index.
+    /// Works similarly to `split_at` in the core library.
+    pub fn split_at(&self, mid: usize) -> (Self, Self) {
+        match self {
+            Self::Empty => (Self::Empty, Self::Empty),
+            Self::Single(chunk) => {
+                if chunk.len() < mid {
+                    return (self.clone(), Self::Empty);
+                }
+                (Self::Single(&chunk[..mid]), Self::Single(&chunk[mid..]))
+            }
+            Self::Multiple { chunks, start, end } => {
+                let mut size = 0usize;
+                for (i, chunk) in chunks.iter().enumerate() {
+                    if size + chunk.len() >= (mid + start) {
+                        return (
+                            Self::new_with_cursors(&chunks[..=i], *start, start + mid),
+                            Self::new_with_cursors(&chunks[i..], mid + start - size, end - size),
+                        );
+                    }
+                    size += chunk.len();
+                }
+                (self.clone(), Self::Empty)
+            }
+        }
+    }
+
+    /// Returns the length of the borrowed payload
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Empty => 0,
+            Self::Single(chunk) => chunk.len(),
+            Self::Multiple { start, end, .. } => end - start,
+        }
+    }
+
+    /// Flatten the slice of byte slices to an owned vector of bytes
+    pub fn to_vec(&self) -> Vec<u8> {
+        let mut vec = Vec::with_capacity(self.len());
+        self.copy_to_vec(&mut vec);
+        vec
+    }
+
+    /// Returns true if the payload is empty
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Copy so many bytes from the payload
+    pub fn take_up_to(&self, limit: usize) -> Vec<u8> {
+        self.split_at(limit).0.to_vec()
+    }
+
+    /// Borrow so many bytes from the payload
+    pub fn borrow_up_to(&self, limit: usize) -> Self {
+        self.split_at(limit).0
+    }
+
+    fn new_with_cursors(chunks: &'a [&'a [u8]], start: usize, end: usize) -> Self {
+        if end - start == 0 {
+            return Self::Empty;
+        }
+        if chunks.len() == 1 {
+            return Self::Single(&chunks[0][start..end]);
+        }
+        Self::Multiple { chunks, start, end }
+    }
+}
+
+impl<'a> From<&'a [u8]> for BorrowedPlainPayload<'a> {
+    fn from(payload: &'a [u8]) -> Self {
+        Self::new_single(payload)
+    }
+}
+
+impl<'a, const N: usize> From<&'a [u8; N]> for BorrowedPlainPayload<'a> {
+    fn from(payload: &'a [u8; N]) -> Self {
+        Self::new_single(payload)
+    }
+}
+
+impl<'a> From<BorrowedPlainPayload<'a>> for Vec<u8> {
+    fn from(payload: BorrowedPlainPayload<'a>) -> Self {
+        payload.to_vec()
+    }
+}
+
+impl<'a> From<BorrowedPlainPayload<'a>> for Payload<'a> {
```

Comment:
We generally keep `impl`s closer to their implementing type definitions, but I agree that this is more tightly coupled to `BorrowedPlainPayload` than to `Payload`. Maybe reverse this to be an `Into`-only impl (and suppress the lint if necessary)?

Reply:
This `From<BorrowedPlainPayload>` implementation for `Payload` as well as `Vec<u8>` are unused, I propose to simply remove them.